src/Orleans.Runtime/Catalog/ActivationDirectory.cs

```diff
@@ -1,27 +1,46 @@
 using System.Collections;
 using System.Collections.Concurrent;
 using System.Collections.Generic;
+using System.Threading;
 
-namespace Orleans.Runtime
+namespace Orleans.Runtime;
+
+internal sealed class ActivationDirectory : IEnumerable<KeyValuePair<GrainId, IGrainContext>>
 {
-    internal sealed class ActivationDirectory : IEnumerable<KeyValuePair<GrainId, IGrainContext>>
+    private int _activationsCount;
+
+    private readonly ConcurrentDictionary<GrainId, IGrainContext> _activations = new();
+
+    public ActivationDirectory()
     {
-        private readonly ConcurrentDictionary<GrainId, IGrainContext> _activations = new();
+        CatalogInstruments.RegisterActivationCountObserve(() => Count);
+    }
 
-        public int Count => _activations.Count;
+    public int Count => Volatile.Read(ref _activationsCount);
```

Comment:
Volatile is not necessary here, by the way: interlocked operations ensure cache coherence already